subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -275,87 +316,273 @@ func (h *HivedAlgorithm) GetAffinityGroup(name string) api.AffinityGroup {
+	for vcn, vcReservation := range reservedCells {
 		for rid, physical := range vcReservation {
+			h.numVCFreeCells[vcn][physical.GetChain()][physical.GetLevel()]--
+			h.totalNumVCFreeCells[physical.GetChain()][physical.GetLevel()]--
 			h.removeCellFromFreeList(physical)
-			virtualList := h.vcSchedulers[vc].getReservedCellList()[rid]
+			virtualList := h.vcSchedulers[vcn].getReservedCellList()[rid]
 			virtual := virtualList[CellLevel(len(virtualList))][0].(*VirtualCell)
-			virtual.SetPhysicalCell(physical)
-			physical.SetVirtualCell(virtual)
-			klog.Infof("Cells bound: %v and %v (reservation)", virtual.GetName(), physical.GetName())
+			bindCell(physical, virtual)
+		}
+	}
+}
+
+func (h *HivedAlgorithm) setBadNode(nodeName string) {
+	if h.badNodes.Contains(nodeName) {
+		return
+	}
+	h.badNodes.Add(nodeName)
+	for _, ccl := range h.fullCellList {
+		for _, gpu := range ccl[1] {
+			pGpu := gpu.(*PhysicalCell)
+			nodes, _ := pGpu.GetPhysicalPlacement()
+			if nodes[0] == nodeName {
+				h.setBadCell(pGpu)
+			}
+		}
+	}
+}
+
+func (h *HivedAlgorithm) setHealthyNode(nodeName string) {
+	if !h.badNodes.Contains(nodeName) {
+		return
+	}
+	h.badNodes.Delete(nodeName)
+	for _, ccl := range h.fullCellList {
+		for _, gpu := range ccl[1] {
+			pGpu := gpu.(*PhysicalCell)
+			nodes, _ := pGpu.GetPhysicalPlacement()
+			if nodes[0] == nodeName {
+				h.setHealthyCell(pGpu)
+			}
+		}
+	}
+}
+
+// setBadCell marks a physical cell as bad and recursively for its parent, guaranteeing that
+// a cell is bad if all of its children are bad. The virtual cell that this physical cell
+// is bound to will also be marked as bad. If the cell has not been bound to a virtual cell,
+// but the scheduler finds that a VC will be allocated such a bad cell,
+// one of the cell in that VC will be marked as bad (the first free cell of this type in the VC).
+func (h *HivedAlgorithm) setBadCell(c *PhysicalCell) {
+	c.GetAPIStatus().CellHealthiness = api.CellBad
+	if c.virtualCell != nil {
+		c.GetAPIStatus().VirtualCell.CellHealthiness = api.CellBad
+		c.virtualCell.GetAPIStatus().CellHealthiness = api.CellBad
+		c.virtualCell.GetAPIStatus().PhysicalCell.CellHealthiness = api.CellBad
+	}
+	chain, level := c.GetChain(), c.GetLevel()
+	h.numBadCells[chain][level]++
+	if h.totalNumVCFreeCells[chain][level] > h.numLeftCells[chain][level]-h.numBadCells[chain][level] {
+		klog.Warningf("Cell type %v (chain %v level %v) now has fewer healthy cells (healthy %v, bad %v) "+
+			"than the total free cells of all the VCs (%v). Certain VCs may be allocated bad cells.",
+			h.cellTypes[chain][level], chain, level, h.numLeftCells[chain][level]-h.numBadCells[chain][level],
+			h.numBadCells[chain][level], h.totalNumVCFreeCells[chain][level])
+		for vcName, vcQuota := range h.numVCFreeCells {
+			if vcQuota[chain][level] > h.numLeftCells[chain][level]-h.numBadCells[chain][level] {
+				klog.Warningf("Cell type %v (chain %v level %v) now has fewer healthy cells (healthy %v, bad %v) "+
```

Comment:
Seems it just extend Yifan's logic to more fine-grain. And we do not really need to do simulation, right?

If so, why previous you say it is very hard to directly calculate?

Reply:
Because I just figured out how to calculate the number of available and bad cells...